rir/src/interpreter/interp.cpp

```diff
@@ -475,286 +526,312 @@ INLINE SEXP fixupAST(SEXP call, Context* ctx, size_t nargs) {
     }
 
-    argslist_ = argslist;
-    validArgslist_ = true;
+    SEXP actuals = createLegacyLazyArgsList(call, ctx);
+    PROTECT(actuals);
+
+    // Patch the already evaluated object into the first entry of
+    // the promise args list
+    SET_PRVALUE(CAR(actuals), obj);
+    SEXP result = dispatchApply(call.popArgs(ctx), obj, actuals, ctx);
+    UNPROTECT(1);
+    return result;
 }
 
-void EnvironmentProxy::init() {
-    if (validREnv_)
-        return;
-    if (!createEnvironment_)
-        return;
+SEXP dispatchApply(const FinalDispatchContext& call, SEXP obj, SEXP actuals,
+                   Context* ctx) {
+    CallSite* cs = call.callSite;
+    SEXP ast = call.ast;
+    SEXP selector = cp_pool_at(ctx, *cs->selector());
+    SEXP op = SYMVALUE(selector);
 
-    SEXP argslist = ctxt_.ap->argslist(parent_);
-    Protect p(argslist);
-    SEXP env = closureArgumentAdaptor(ctxt_.call, ctxt_.callee, argslist,
-                                      parent_->env(), R_NilValue);
+    // ===============================================
+    // First try S4
+    if (IS_S4_OBJECT(obj) && R_has_methods(op)) {
+        SEXP result =
+            R_possible_dispatch(ast, op, actuals, call.callerEnv, TRUE);
+        if (result)
+            return result;
+    }
 
-    // This patches the missing values in the return context of this
-    // closure call. From now, they are also protected during gc
-    // (contexts are considered roots by the collector).
-    RCNTXT* cntxt = findClosureReturnContext();
-    cntxt->promargs = argslist;
-    cntxt->cloenv = env;
+    // ===============================================
+    // Then try S3
+    const char* generic = CHAR(PRINTNAME(selector));
+    SEXP rho1 = Rf_NewEnvironment(R_NilValue, R_NilValue, call.callerEnv);
+    ostack_push(ctx, rho1);
+    RCNTXT cntxt;
+    initClosureContext(call, &cntxt, rho1, call.callerEnv, actuals, op);
+    SEXP result;
+    bool success = Rf_usemethod(generic, obj, ast, actuals, rho1,
+                                call.callerEnv, R_BaseEnv, &result);
+    ostack_pop(ctx);
+    endClosureContext(&cntxt, success ? result : R_NilValue);
+    if (success)
+        return result;
+
+    // ===============================================
+    // Now normal dispatch
+    auto fcall = call.setCallee(Rf_findFun(selector, call.callerEnv));
+
+    // TODO something should happen here
+    if (fcall.callee == R_UnboundValue)
+        assert(false && "Unbound var");
+    if (fcall.callee == R_MissingArg)
+        assert(false && "Missing argument");
+
+    switch (TYPEOF(fcall.callee)) {
+    case SPECIALSXP:
+        return legacySpecialCall(fcall, ctx);
+    case BUILTINSXP:
+        // force all promises in the args list. We created a lazy legacy args
+        // list, since we did not know the callee yet. So now we know and need
+        // to make it eager.
+        for (SEXP a = actuals; a != R_NilValue; a = CDR(a))
+            SETCAR(a, Rf_eval(CAR(a), call.callerEnv));
```

Comment:
This `eval` call just forces the promise, not? Why not force it ourselves here?
Or even `Rf_evalList(actuals, call.callerEnv, call.ast, 0);`?